### DISCUSSION TITLE
How is that attempt to get AcidBox on one core of an ESP32 S3 and aciduino running on the other core?

### DISCUSSION THREAD

**Discussion by 01GOD** [copych/AcidBox#33] [General]
Hi! 

Saw some posts in the Aciduino repo. How is that attempt going?

**Comment by copych** [comment]
Haven't returned to that topic yet. Been busy refactoring envelopes. Now gonna tune them a bit better.

**Comment by 01GOD** [comment]
Oh, nice. Added a separate accent envelope to make tuning accented notes
envelope easier?

**Reply by copych** [reply]
Not exactly), just made two classes instead of those two functions, now it uses arithmetics instead of lookup tables, a bit more flexible, though there's not much of performance increase. Also tested simple oscillator instead of LUTs, but it has IF for a square wave, so again similar performance. Gonna try oversampling to eliminate aliasing, hopefully there're some cpu power left.

**Comment by 01GOD** [comment]
Ah, but an accurate emulation of that square wave isn’t a normal square
wave. Looks more like a LOTUS PETAL shape. Thus, making that with a lookup
table is maybe best digitally.

**Reply by copych** [reply]
I know, Anton said it should look like this
![image](https://github.com/user-attachments/assets/c6bf967b-a638-41b5-ba05-47ae215f7acb)
My variants are 'slightly' cleaner, I have seen these on some h/w forum, but someday we can try Anton's waveforms and his approach.
my sqare:
![image](https://github.com/user-attachments/assets/9944d5a4-1d36-4548-9b70-e566bfc1b7bb)
my saw:
![image](https://github.com/user-attachments/assets/d3955b71-6367-46c8-8b23-83bbd78c5e44)

**Reply by 01GOD** [reply]
That exponential style saw looks alright, and obviously makes the transient sound louder. 

That square wave doesn't look as passable though..."one of these pictures is not like the other"

**Reply by 01GOD** [reply]
Could put 2 of those saw waves back to back on top of a standard square wave for a more accurate vague approximation.

**Comment by copych** [comment]
The matter is that absolutely differently looking waveforms may sound quite the same because of the same frequency contents. Humans don't hear phase differences.
Very good illustration of this fact https://youtu.be/Ffka-hPzug0?feature=shared

**Comment by copych** [comment]
There's a branch https://github.com/copych/AcidBox/tree/S3-only with the latest modifications. If you have s3, I invite you for testing.
Well, if you don't have s3 I still suggest testing, cause after optimizing s3 timing shows that even running on just one core, there's a place for one more 303, so maybe esp32 would do the thing.

**Comment by 01GOD** [comment]
Saw now. Thanks for notifying!👍🏻 I actually already made a majorly
modified edition that mainly runs on one core. That is with only one synrh
and drum machine though.

Would like to try that with Aciduino, so may install it sometime soon to
try it. I made a custom interface, so it would involve adapting it to that.
May try it with MIDI from a DAW sooner though.

Aciduino on ESP32 is direct USB MIDI, witnout any serial bridge app
involved, right?

**Reply by copych** [reply]
> Aciduino on ESP32 is direct USB MIDI, witnout any serial bridge app
involved, right?

Only S2 and S3 has native USB support, but S2 doesn't have FPU, so again only S3 will do.

**Reply by 01GOD** [reply]
Was at a Maker Faire this past weekend, so excuse the delayed replies. 

Right, I meant ESP32 S3. Glad to hear that is native USB MIDI, instead of through a serial bridge!

**Reply by 01GOD** [reply]
Tried plugging in a USB MIDI keyboard directly already?

**Reply by copych** [reply]
Yes, I have tried to use ESP32S2 and ESP32S3 as a USB host to connect a USB midi keyboard, and I failed that attempt. I started an issue at Espressif repo, and there were some updetes since then, but I never succeed. I even thougnt that my kbd isn't good enough (has some complicated inner structure) to go with MCUs, but RP2040 recognizes it and runs smoothly, so it's not the case. Last time I have tried ESP USB HOST was about 3 months ago, so maybe it's time to try again...

**Reply by andyvans** [reply]
Here is an example of using a keyboard https://github.com/tanakamasayuki/EspUsbHost.

You also need to solder the pad on the underside of the ESP32S3 units that you get off AliExpress. This enables OTG. Board = esp32s3usbotg

**Comment by copych** [comment]
Thank you for the link, I will try this as well as esp-idf usb host example.
PS I also try to connect a mouse to the board, and it is recognized well. So it's only an issue with the kbd

**Comment by 01GOD** [comment]
@copych: MERRY CHRISTMAS!

**Reply by copych** [reply]
@01GOD thank you!
